service/matching/matching_engine.go

```diff
@@ -1261,7 +1261,69 @@ func (e *matchingEngineImpl) DescribeTaskQueue(
 	if err != nil {
 		return nil, err
 	}
-	return pm.LegacyDescribeTaskQueue(req.GetIncludeTaskQueueStatus())
+	descrResp, err := pm.LegacyDescribeTaskQueue(req.GetIncludeTaskQueueStatus())
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: add cache
+	if req.ReportStats {
+		totalStats := &taskqueuepb.TaskQueueStats{}
+
+		// find number of partitions
+		rootPartition, err := tqid.PartitionFromProto(req.GetTaskQueue(), request.GetNamespaceId(), req.GetTaskQueueType())
+		if err != nil {
+			return nil, err
+		}
+		tqConfig := newTaskQueueConfig(rootPartition.TaskQueue(), e.config, namespace.Name(req.Namespace))
+		numPartitions := max(tqConfig.NumWritePartitions(), tqConfig.NumReadPartitions())
+
+		// find all buildIds
+		var buildIds []string
+		userData, _, err := pm.GetUserDataManager().GetUserData()
+		if err != nil {
+			return nil, err
+		}
+		if userData != nil {
+			if typedUserData := userData.GetData().GetPerType()[int32(pm.Partition().TaskType())]; typedUserData != nil {
+				for _, v := range typedUserData.GetDeploymentData().GetVersions() {
+					buildIds = append(buildIds, v.GetVersion().GetBuildId())
```

Comment:
I'm not sure but there may be "version"s without build id or deployment name stored here for historical reasons, it's nice to guard agains them (they could be ignored for this purpose).

But more importantly, the V3 versions are identified by combination of deployment name + build ID. So to make it work, you can either pass `worker_versioning.WorkerDeploymentVersionToString(v.GetVersion())` receiving side will handle it ok.

Or if you want to be fancier, you can add a `repeated temporal.api.deployment.v1.WorkerDeploymentVersion versions = 4;` in `TaskQueueVersionSelection` to properly pass the version object. But this is something we'd do later anyways with other placed needing to get rid of the string representation for version in the protos.